### PR TITLE
PCHR-2164: Files tab component and bug fixes

### PR DESCRIPTION
**Pull request by deb1990**
## Overview
As part of this PR(and https://github.com/compucorp/civihr-employee-portal-theme/pull/186) a new component for the files tab of the leave request popup has been created. Also It includes some bug fixes

## Problem 1
Long file's name overlaps the Remove Link
**Before**
manager_leave___hr17-1

**Solution**
Added `chr_leave-request-files_file-name` class to clip the filename after a certain length.

**After**
my_leave___hr17-1

## Problem 2
Save button is enabled for TOIL requests regardless of changes were made or not
**Before**
2164b

**Solution**
Added `parentRequestCtrl.canSubmit` call in the `canSubmit` of `toil-request-ctrl`.

**After**
2164a

## Files tab Component

A new component called `leaveRequestPopupFilesTab` has been created and the code, unit tests related to that is moved.

**How to use**
`<leave-request-popup-files-tab request="<>" mode="<>" can-manage="<>"></leave-request-popup-files-tab>`

**Attributes**
`request` - Object - The leave request object
`mode` - String - The mode, i.e. view/edit/create.
`can-manage` - Boolean - If role is admin or manager

## Downloadable Attachments
The attachments are now downloadable, when clicked on already uploaded attachments, the same will be downloaded in a new tab.
Note that in staff/manager mode the download is not working as there is permission issue(PCHR-2398 is created).
2164d

## Comments
Before refactoring, the files tab was using `getCommentorName` function to get the contact name of attachment author. It was wrong because
1. Files should not use the contacts fetched by `comments`. 
2. The Attachment API does not support saving the contact id so there is no way we can show the contact name for already saved attachments. This is a known Issue as civicrm core needs to be changed.

So `getAuthorName()` is created in files tab components, which returns "Me" if the comment is not saved yet. Added to todo to change to logic after Attachment API supports the contact id.

---

- [x] Tests Pass

### CODE REVIEW

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-request-popup-files-tab.js [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+
+define([
+  'common/lodash',
+  'common/moment',
+  'leave-absences/shared/modules/components',
+  'common/services/hr-settings'
+], function (_, moment, components) {
+  components.component('leaveRequestPopupFilesTab', {
+    bindings: {
+      // canManage: '<',
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-request-popup-files-tab.js]: commented out code

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-request-popup-files-tab.js [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+      mode: '<',
+      request: '<'
+    },
+    templateUrl: ['shared-settings', function (sharedSettings) {
+      return sharedSettings.sharedPathTpl + 'directives/leave-request-popup/leave-request-popup-files-tab.html';
+    }],
+    controllerAs: 'vm',
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-request-popup-files-tab.js]: give a meaningful name to the controller alias

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-request-popup-files-tab.js [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+    $log.debug('Component: leave-request-popup-files-tab');
+
+    var filesContacts = [];
+    var vm = Object.create(this);
+    vm.supportedFileTypes = '';
+    vm.today = Date.now();
+    vm.userDateFormatWithTime = HRSettings.DATE_FORMAT + ' HH:mm';
+    vm.userDateFormat = HRSettings.DATE_FORMAT;
+
+    (function init () {
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/components/leave-request-popup-files-tab.js]: the `init` function should be between the public (`vm.xyz`) and private (`function xyz`) methods, like for any other component

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/components/leave-request-popup-files-tab.spec.js [hunk 4]
```diff
@@ -0,0 +1,66 @@
+/* eslint-env amd, jasmine */
+/* global CRM, inject */
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/components/leave-request-popup-files-tab.spec.js]: unnecessary

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/components/leave-request-popup-files-tab.spec.js [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+    });
+
+    describe('when user uploads files', function () {
+      describe('when number of files are below allowed limit', function () {
+        beforeEach(function () {
+          controller.request.files = [1, 2, 3, 4];
+          controller.request.fileUploader.queue = [1, 2];
+        });
+
+        it('returns true', function () {
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/components/leave-request-popup-files-tab.spec.js]: this is what it reads: *"when user uploads files / when number of files are below allowed limit / it returns true"*

the subject is missing

File: uk.co.compucorp.civicrm.hrleaveandabsences/views/shared/directives/leave-request-popup/leave-request-popup.html [hunk 6]
```diff
@@ -66,9 +66,12 @@
               </leave-request-popup-comments-tab>
             </uib-tab>
             <uib-tab heading="Files">
-              <ng-include
-                src="sharedPathTpl + 'directives/leave-request-popup/leave-request-popup-files-tab.html'">
-              </ng-include>
+              <leave-request-popup-files-tab
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/views/shared/directives/leave-request-popup/leave-request-popup.html]: i think everything in one line reads better, it's not that long anyway
```html
<leave-request-popup-files-tab request="$ctrl.request" mode="$ctrl.mode" can-manage="$ctrl.canManage"></leave-request-popup-files-tab>
```

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/sub-controllers/toil-request-ctrl.js [hunk 7]
```diff
@@ -70,7 +70,7 @@ define([
        * @return {Boolean}
        */
       vm.canSubmit = function () {
-        return !!vm.request.toil_duration && !!vm.request.toil_to_accrue &&
+        return parentRequestCtrl.canSubmit.call(this) && !!vm.request.toil_duration && !!vm.request.toil_to_accrue &&
```
Comment (AkA84) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/src/leave-absences/shared/controllers/sub-controllers/toil-request-ctrl.js]: maybe it's better, for readability-sake, to put each condition in a new line

File: uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/components/leave-request-popup-files-tab.spec.js (hunk 5 above)
Comment (deb1990) [inline uk.co.compucorp.civicrm.hrleaveandabsences/js/angular/test/shared/components/leave-request-popup-files-tab.spec.js]: Added `canUploadMore()` in the first describe.

**Pull request merged**